nixos/modules/services/networking/teleport.nix

```diff
@@ -0,0 +1,76 @@
+{ config, pkgs, lib, ... }:
+
+with lib;
+
+let
+  cfg = config.services.teleport;
+
+  # Pretty-print JSON to a file
+  writePrettyJSON = name: x:
+    pkgs.runCommand name { } ''
+      echo '${builtins.toJSON x}' | ${pkgs.jq}/bin/jq . > $out
+    '';
+  # This becomes the main config file
+  # To be used in the future when I add seperate option sections for each part of teleport
+  # Rather then writing a single yaml file directly into your nix file.
+  telConfig = {
+    global = cfg.globalConfig;
+    rule_files = cfg.ruleFiles ++ [
+      (pkgs.writeText "teleport.rules" (concatStringsSep "\n" cfg.rules))
+    ];
+  };
+
+  generatedTeleportYml = writePrettyJSON "teleport.yaml" telConfig;
+
+  teleportYml =
+    if cfg.configText != null then
+      pkgs.writeText "teleport.yml" cfg.configText
+    else generatedTeleportYml;
+
+  cmdlineArgs = "-c ${teleportYml}";
+
+in
+
+{
+  options = {
+    services.teleport = {
+      enable = mkOption {
+        type = types.bool;
+        default = false;
+        description = ''
+          Enable the Teleport Cluster SSH daemon.
+        '';
+      };
+      dataDir = mkOption {
+        type = types.path;
+        default = "/var/lib/teleport/";
+        description = ''
+          Directory to store Teleport Service data.
+        '';
+      };
+      configText = mkOption {
```

Comment:
Having a single `settings` option to represent this file might be worth looking into and match upstream more closely. What were aiming for by splitting this file into multiple? I'm not familiar with this program so don't know the pros and cons of doing so.

Reply:
I'm not certain what lines you're referring to here? Where am I splitting the file into multiple?

Reply:
I was mentioning the note you made above:
```
  # This becomes the main config file
  # To be used in the future when I add seperate option sections for each part of teleport
  # Rather then writing a single yaml file directly into your nix file.
```

Since the file format is `yaml` this maps nicely to a `nix` attribute set. Very shortly `nix` will have even nicer support for this, see https://github.com/NixOS/nixpkgs/pull/75584 if interested.

Reply:
Ah my apologies, to be clear what I actually meant, was what you implied. Having sub services as attr sets that can be written to the tank config file.

Rather then just encoding a flat yaml file directly in your nix expression when you enable this service.

Reply:
This should now make more sense, with the changes I've made to define the service config file in nix expr lang.